pubsub/solace/amqp/amqp.go

```diff
@@ -0,0 +1,289 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package amqp
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	time "time"
+
+	amqp "github.com/Azure/go-amqp"
+
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	publishRetryWaitSeconds = 2
+	publishMaxRetries       = 3
+)
+
+// amqpPubSub type allows sending and receiving data to/from an AMQP 1.0 broker
+type amqpPubSub struct {
+	session           *amqp.Session
+	metadata          *metadata
+	logger            logger.Logger
+	publishLock       sync.RWMutex
+	publishRetryCount int
+	ctx               context.Context
+	cancel            context.CancelFunc
+}
+
+// NewAMQPPubsub returns a new AMQPPubSub instance
+func NewAMQPPubsub(logger logger.Logger) pubsub.PubSub {
+	return &amqpPubSub{
+		logger:      logger,
+		publishLock: sync.RWMutex{},
+	}
+}
+
+// Init parses the metadata and creates a new Pub Sub Client.
+func (a *amqpPubSub) Init(metadata pubsub.Metadata) error {
+	amqpMeta, err := parseAMQPMetaData(metadata, a.logger)
+	if err != nil {
+		return err
+	}
+
+	a.metadata = amqpMeta
+
+	a.ctx, a.cancel = context.WithCancel(context.Background())
+
+	s, err := a.connect()
+	if err != nil {
+		return err
+	}
+
+	a.session = s
+
+	return err
+}
+
+func AddPrefixToAddress(t string) string {
+	dest := t
+
+	// Unless the request comes in to publish on a queue, publish directly on a topic
+	if !strings.HasPrefix(dest, "queue:") && !strings.HasPrefix(dest, "topic:") {
+		dest = "topic://" + dest
+	} else if strings.HasPrefix(dest, "queue:") {
+		dest = strings.Replace(dest, "queue:", "queue://", 1)
+	} else if strings.HasPrefix(dest, "topic:") {
+		dest = strings.Replace(dest, "topic:", "topic://", 1)
+	}
+
+	return dest
+}
+
+// Publish the topic to amqp pubsub
+func (a *amqpPubSub) Publish(req *pubsub.PublishRequest) error {
+	a.publishLock.Lock()
+	defer a.publishLock.Unlock()
+
+	a.publishRetryCount = 0
+
+	if req.Topic == "" {
+		return errors.New("topic name is empty")
+	}
+
+	m := amqp.NewMessage(req.Data)
+
+	// If the request has ttl specified, put it on the message header
+	ttlProp := req.Metadata["ttlInSeconds"]
+	if ttlProp != "" {
+		ttlInSeconds, err := strconv.Atoi(ttlProp)
+		if err != nil {
+			a.logger.Warnf("Invalid ttl received from message %s", ttlInSeconds)
+		} else {
+			m.Header.TTL = time.Second * time.Duration(ttlInSeconds)
+		}
+	}
+
+	sender, err := a.session.NewSender(
+		amqp.LinkTargetAddress(AddPrefixToAddress(req.Topic)),
+	)
+
+	if err != nil {
+		a.logger.Errorf("Unable to create link to %s", req.Topic, err)
+	} else {
+		err = sender.Send(a.ctx, m)
+
+		// If the publish operation has failed, attempt to republish a maximum number of times
+		// before giving up
+		if err != nil {
+			for a.publishRetryCount <= publishMaxRetries {
+				a.publishRetryCount++
```

Comment:
I think as per tracking issue https://github.com/dapr/components-contrib/issues/1797, we need to disable built-in re-tries for publishing as well. So, may be in new components, it may not be a recommendable practice to include this.
Thoughts @berndverst @ItalyPaleAle?

Reply:
We can keep this for now to unblock this. There will be more components which need updating and  the strategy for this consistently isn't totally clear yet @DeepanshuA.

Reply:
Sure, thanks for the clarification @berndverst